Improve test coverage

tpdownes opened the pull request:
### Submission Checklist

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cloud HPC Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

mr0re1 commented on pkg/config/expand.go:
```diff
@@ -585,3 +585,45 @@ func (dg DeploymentGroup) OutputNames() []string {
+	if thisGroupIdx == -1 {
+		return nil, fmt.Errorf("this group wasn't found in the deployment metadata")
+	}
+	if thisGroupIdx == 0 {
```
It's inconsequential, nothing changes if you remove it (except returned nil-map that can be a nuisance), lets remove it.

mr0re1 commented on pkg/config/expand.go:
```diff
@@ -585,3 +585,45 @@ func (dg DeploymentGroup) OutputNames() []string {
+
+// GetIntergroupOutputNamesByGroup returns a map from group names to a list of
+// outputs that are needed by this group
+func GetIntergroupOutputNamesByGroup(thisGroup GroupName, dc DeploymentConfig) (map[GroupName][]string, error) {
```
nit. s/thisGroup/n/

mr0re1 replied:
nit. Let's make it a function of `DeploymentGroup` to remove all that "index of" lines, by just inspecting all the groups

mr0re1 replied:
nit. s/GetIntergroupOutputNamesByGroup/InputsByGroup/

mr0re1 commented on pkg/config/expand.go:
```diff
@@ -585,3 +585,45 @@ func (dg DeploymentGroup) OutputNames() []string {
+// GetIntergroupOutputNamesByGroup returns a map from group names to a list of
+// outputs that are needed by this group
+func GetIntergroupOutputNamesByGroup(thisGroup GroupName, dc DeploymentConfig) (map[GroupName][]string, error) {
+	thisGroupIdx := dc.Config.GroupIndex(thisGroup)
```
nit. s/thisGroupIdx/i/

mr0re1 commented on pkg/config/expand.go:
```diff
@@ -585,3 +585,45 @@ func (dg DeploymentGroup) OutputNames() []string {
+		return nil, nil
+	}
+
+	thisIntergroupRefs := dc.Config.DeploymentGroups[thisGroupIdx].FindAllIntergroupReferences(dc.Config)
```
nit. s/thisIntergroupRefs/refs/

mr0re1 commented on pkg/config/expand.go:
```diff
@@ -585,3 +585,45 @@ func (dg DeploymentGroup) OutputNames() []string {
+	}
+
+	thisIntergroupRefs := dc.Config.DeploymentGroups[thisGroupIdx].FindAllIntergroupReferences(dc.Config)
+	thisIntergroupInputNames := make([]string, len(thisIntergroupRefs))
```
nit. s/thisIntergroupInputNames/in/

mr0re1 commented on pkg/config/expand.go:
```diff
@@ -585,3 +585,45 @@ func (dg DeploymentGroup) OutputNames() []string {
+
+// return sorted list of elements common to s1 and s2
+func intersection(s1 []string, s2 []string) []string {
+	count := make(map[string]int)
```
nit. s/count/left or first/ and make it `map[string]bool`

mr0re1 commented on pkg/config/expand.go:
```diff
@@ -585,3 +585,45 @@ func (dg DeploymentGroup) OutputNames() []string {
+
+	foundInBoth := map[string]bool{}
+	for _, v := range s2 {
+		if count[v] > 0 {
```
nit. s/foundInBoth/both/
```go
for _, v := range s2 {
  both[v] = left[v]
}
```

mr0re1 commented on pkg/config/expand_test.go:
```diff
@@ -497,3 +497,23 @@ func (s *MySuite) TestValidateModuleReference(c *C) {
+	is = intersection([]string{"A", "C"}, []string{"D", "C", "B", "A"})
+	c.Assert(is, DeepEquals, []string{"A", "C"})
+
+	is = intersection([]string{"A", "C"}, []string{})
```
nit. You can `s/[]string{}/nil/` for stronger assertion.

mr0re1 commented on pkg/shell/common_test.go:
```diff
@@ -101,3 +83,57 @@ func (s *MySuite) TestCheckWritableDir(c *C) {
+	m2 := map[string]int{}
+
+	m1["foo"] = 0
+	m2["bar"] = 0
```
nit. Lets one-line `m1` & `m2` initializations.

mr0re1 reviewed (approved):
LGTM. All comments are `nit`

tpdownes committed:
Improve test coverage

- where appropriate, migrate functionality from pkg/shell to pkg/config
- eliminate unnecessary functions in favor of pre-existing functions
- add tests for untested functions
